instance/models/instance.py

```diff
@@ -0,0 +1,476 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - Instance
+"""
+
+# Imports #####################################################################
+
+import os
+
+from django.conf import settings
+from django.core.validators import RegexValidator
+from django.db import models
+from django.template import loader
+from django.utils import timezone
+from django_extensions.db.models import TimeStampedModel
+
+from instance import ansible, github
+from instance.gandi import GandiAPI
+from instance.github import fork_name2tuple, get_username_list_from_team
+from instance.log_exception import log_exception
+from instance.repo import clone_configuration_repo
+from instance.models.logging_mixin import LoggerInstanceMixin
+from instance.models.utils import ValidateModelMixin
+
+
+# Constants ###################################################################
+
+PROTOCOL_CHOICES = (
+    ('http', 'HTTP - Unencrypted clear text'),
+    ('https', 'HTTPS - Encrypted'),
+)
+
+gandi = GandiAPI()
+
+
+# Exceptions ##################################################################
+
+class InconsistentInstanceState(Exception):
+    """
+    Indicates that the status of an instance can't be determined
+    """
+    pass
+
+
+# Validators ##################################################################
+
+sha1_validator = RegexValidator(regex='^[0-9a-f]{40}$', message='Full SHA1 hash required')
+
+
+# Models ######################################################################
+
+class Instance(ValidateModelMixin, TimeStampedModel):
+    """
+    Instance - Group of servers running an application made of multiple services
+    """
+    EMPTY = 'empty'
+    NEW = 'new'
+    STARTED = 'started'
+    ACTIVE = 'active'
+    BOOTED = 'booted'
+    PROVISIONED = 'provisioned'
+    REBOOTING = 'rebooting'
+    READY = 'ready'
+    LIVE = 'live'
+    STOPPING = 'stopping'
+    STOPPED = 'stopped'
+    TERMINATING = 'terminating'
```

Comment:
You could create a new file status.py which contains Status, InstanceStatus, and ServerStatus definitions / state machines.... but that's perhaps getting complicated.

The minimum I would suggest you change for sure is adding some documentation to the constants above. They're documented in server.py but for someone who doesn't know that, it's very difficult to know things like what the difference between "BOOTED" and "STARTED" is or between "ACTIVE", "READY", and "LIVE" is.